src/access.rs

```diff
@@ -0,0 +1,438 @@
+use bdk::{SignOptions, SyncOptions};
+
+use bitcoin::{BlockHash, BlockHeader, Script, Transaction, Txid};
+
+use std::collections::HashSet;
+use std::sync::{Arc, Mutex};
+
+pub struct LdkLiteChainAccess<D>
+where
+	D: BatchDatabase,
+{
+	// A BDK blockchain used for wallet sync.
+	blockchain: EsploraBlockchain,
+	// A BDK on-chain wallet.
+	wallet: Mutex<bdk::Wallet<D>>,
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	queued_transactions: Mutex<HashSet<Txid>>,
+	// Transactions that were previously processed, but must not be forgotten yet.
+	watched_transactions: Mutex<HashSet<Txid>>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	queued_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Outputs that were previously processed, but must not be forgotten yet.
+	watched_outputs: Mutex<HashSet<WatchedOutput>>,
+	// The tip hash observed during our last sync.
+	last_sync_hash: tokio::sync::Mutex<Option<BlockHash>>,
+	logger: Arc<FilesystemLogger>,
+}
+
+impl<D> LdkLiteChainAccess<D>
+where
+	D: BatchDatabase,
+{
+	pub(crate) fn new(
+		blockchain: EsploraBlockchain, wallet: bdk::Wallet<D>, logger: Arc<FilesystemLogger>,
+	) -> Self {
+		let wallet = Mutex::new(wallet);
+		let watched_transactions = Mutex::new(HashSet::new());
+		let queued_transactions = Mutex::new(HashSet::new());
+		let watched_outputs = Mutex::new(HashSet::new());
+		let queued_outputs = Mutex::new(HashSet::new());
+		let last_sync_hash = tokio::sync::Mutex::new(None);
+		Self {
+			blockchain,
+			wallet,
+			queued_transactions,
+			watched_transactions,
+			queued_outputs,
+			watched_outputs,
+			last_sync_hash,
+			logger,
+		}
+	}
+
+	pub(crate) async fn sync_wallet(&self) -> Result<(), Error> {
+		let sync_options = SyncOptions { progress: None };
+
+		self.wallet
+			.lock()
+			.unwrap()
+			.sync(&self.blockchain, sync_options)
+			.map_err(|e| Error::Bdk(e))?
+	}
+
+	/// Syncs the LDK wallet via the `Confirm` interface. We run in a loop until we completed a
+	/// full iteration without
+	pub(crate) async fn sync(&self, confirmables: Vec<&(dyn Confirm + Sync)>) -> Result<(), Error> {
+		// This lock makes sure we're syncing once at a time.
+		let mut locked_last_sync_hash = self.last_sync_hash.lock().await;
+
+		let client = &*self.blockchain;
+
+		let mut tip_hash = client.get_tip_hash().await?;
+
+		loop {
+			let pending_registrations = self.process_queues();
+			let new_tip = Some(tip_hash) != *locked_last_sync_hash;
+
+			// We loop until any registered transactions have been processed at least once, or the
+			// tip hasn't been updated during the last iteration.
+			if pending_registrations || new_tip {
+				// First check for any unconfirmed tranasactions and act on it immediately.
+				self.sync_unconfirmed_transactions(&confirmables).await?;
+
+				// Update the known tip to the newest one.
+				if new_tip {
+					*locked_last_sync_hash = Some(tip_hash);
```

Comment:
Good point, now moved it down so that we only store the tip hash after a successful iteration. Are we fine though with potentially calling `best_block_updated` with the same hash twice in edge cases, or do I need another variable to track what I gave to it previously?